README.md
Safely manage document.write() with events and functions

document.write() is still widely used in Javascript, especially by advertisers. This causes several problems:

* can only write where the script is loaded from
* can't edit text before writing
* can't cancel writing
* can't trigger events on writing
* calls after the ready event will reset the page

To work around these issues, the docwrite plugin:

* provides beforedocwrite and afterdocwrite events around document.write()
* allows document.write() to be edited, redirected and even cancelled altogether
* automatically redirects late document.write() calls to an invisible <div>
* provides $(element).docwrite() that redirects document.write()s in a specified function

You can use the event-based interface like this:

```js
$(document).bind('beforedocwrite', function(event, data) {
    // redirect the write:
    data.target = $('#my_element');
    // Edit what will be written:
    if ( /pattern/.test( data.html ) ) {
        data.html = "you can change the text <i>however you like</i>";
    };
    // Cancel writing:
    if ( !confirm("Should we write this:\n" + data.html) )
        event.preventDefault();
});
$(document).bind('afterdocwrite', function(event, data) {
    alert(data.html + " was written to " + data.target);
});
```

You can use the functional interface like this:

```js
$(element).docwrite( function(arg1, arg2) {
    document.write("my <b>html</b> will be appended to $(element)");
}, "passed to function as arg1",
   "passed to function as arg2"
);
$(element).docwrite(
    "<div>Appended to element</div>\n" +
    "<script type='text/javascript'>" +
        "document.write('nested document.write()s are handled correctly');" +
    "</script>"
);
```

The functional interface doesn't trigger docwrite events:

```js
$(document).bind('afterdocwrite',
    function(event, data) { alert('event triggered'); }
);
$(window).load(function() {
    document.write("will trigger an event (and be appended to a hidden <div>)");
});
$(element).docwrite( "will not trigger an event" );
```

1.1.0. Added ability to process splitted nesteds scripts

```js
$(element).docwrite("<script type='text/javascript' ");
$(element).docwrite("src='alert.js'></scr"+"ipt>");

$(element).docwrite("<div>Appended to element</div>" + "<script type='text/javascript'>" + "m='<div>nested '; document.write(m+'");
$(element).docwrite("document.write()s are handled");
$(element).docwrite(" correctly</div>');" + "</scr"+"ipt><div>");
$(element).docwrite('nezhdanchik</div>');
```
 
Initial release from Slando (http://www.slando.com) Development team. (http://plugins.jquery.com/content/docwrite-100)  
Modified by m8rge

changelog.txt
Docwrite

Version 1.2.0
=============
* Reformatted m8rge's changes slightly
* Added unit tests
* Fixed several parsing issues

Version 1.1.0
=============
* Modified by m8rge
* Added ability to process splitted nesteds scripts
* See https://github.com/m8rge/docwrite

Version 1.0.0
=============
* Initial release from the Slando (http://www.slando.com) Development team.


nested_docwrite_test.js
document.write("(nested document.write() loaded from file)");

jquery.docwrite.js
/*!
 * docwrite jQuery plugin v1.2.0
 *
 * Copyright 2011, Slando Ltd. <dev@slando.com>
 * Dual licensed under the MIT or GPL Version 2 licenses.
 * http://jquery.org/license
 *
 * Date: Mon May  9 11:29:29 BST 2011
 */

/**
 * @fileOverview
 * Make document.write() a little bit safer
 *
 * // Redirect document.write() to an element:
 * $(element).docwrite( function(arg1, arg2) {
 *     document.write("appended to element");
 *     document.write("<script>document.write('also appended to element')</script>");
 * }, "passed to function as arg1", "passed to function as arg2" );
 *
 * $(document).load(function() {
 *     document.write("will be appended to an invisible <DIV>");
 * });
 *
 * $(document).bind( 'beforedocwrite', function(event, data) {
 *     data.html.replace("whatever was going to be written",
 *                       "what you would rather write");
 *     data.target = $(destination);
 *     data.unused = "pass data to the afterdocwrite event";
 *     event.preventDefault(); // cancel writing
 * });
 *
 * $(document).bind( 'afterdocwrite', function(event, data) {
 *     if ( data.unused ) {
 *         # ...
 *     };
 * });
 */
(function($) {

     var match_script     = /<script[ \t\n]([^>]*)>/i,
         match_script_src = /^<script[ \t\n]([^>]*[ \t\n])?src[ \t\n]*=[ \t\n]*/i,
         match_script_end = /<\/script[ \t\n]*>/i,
         document_write   = document.write,
         write_target     = undefined,
         script_buffer    = "";

     // document.write() implementation before the ready event has fired:
     function write(html) {

         var data  = { target: null, html: html },
             event = $.Event("beforedocwrite");

         $(document).trigger( event, data );

         if ( !event.isDefaultPrevented() ) {

             if ( null === data.target )
                 if ( 'call' in document_write ) { // sane browsers
                     document_write.call( document, data.html );
                 } else { // IE
                     // make sure document.write() will be reset, even if we throw an exception:
                     var timeout = setTimeout(function() { document.write = write; }, 0);
                     document.write = document_write;
                     document.write( data.html );
                     document.write = write;
                 }
             else if ( data.target.docwrite )
                 data.target .docwrite( data.html );
             else
                 $(data.target).docwrite( data.html );

             $(document).trigger( 'afterdocwrite', data );

         }

     };
     function writeln(html) { write( html + "\n" ); };

     // document.write() implementation after the ready event has fired:
     function write_late(html) {

         if ( write_target === undefined ) {
             write_target = document.createElement('DIV');
             write_target.style.display = "none";
             document.body.appendChild(write_target);
             write_target = $(write_target);
         };

         var data  = { target: write_target, html: html },
             event = $.Event("beforedocwrite");

         $(document).trigger( event, data );

         if ( !event.isDefaultPrevented() ) {
             data.target.docwrite( data.html );
             $(document).trigger( 'afterdocwrite', data );
         };

     };
     function writeln_late(html) { write_late( html + "\n" ); };

     document.write   = write;
     document.writeln = writeln;
     $(document).ready(function() {
         document.write   = write_late;
         document.writeln = writeln_late;
     });

     $.fn.docwrite = function(arg) {

         var html = [],
             // we're about to override the system document.write (again)
             // but first we make sure it will be reset, even if we throw an exception:
             document_write   = document.write,
             document_writeln = document.writeln,
             timeout = setTimeout(function() {
                 document.write   = document_write;
                 document.writeln = document_writeln;
             }, 0),
             i=0, args=[];

         /*
          * Protect against nested document.write's:
          *
          *     document.write('<script type="text/javascript">document.write("...")</script>');
          *
          * Browsers ignore <script>s inserted with innerHTML,
          * so we have to parse out and eval() the nested <script>s.
          *
          */
         document.write = function write_inner(text) {

//             console.log("have text:",text);
             var start_pos,
                 end_pos; // needs to be a local variable because this is called recursively

             text = script_buffer + text;
             script_buffer = "";

             while ( -1 != ( start_pos = text.search(match_script) ) ) {
                 // <script...>

                 html.push( text.substr(0, start_pos) );

                 text = text.substr( start_pos );

                 if ( match_script_src.test( text ) ) {
                     // <script src="...">: remove the "<script" and continue parsing:
                     html.push( text.substr(0, 7) );
                     text     = text.substr(7   );
                 } else if ( -1 == ( end_pos = text.search(match_script_end) ) ) {
                     // <script> with no </script>:
                     script_buffer = text;
//                     console.log(html, script_buffer);
                     return;
                 } else {
                     // <script>...</script>, may do a document.write() of its own:
                     start_pos = text.indexOf( '>' ) + 1;
                     $.globalEval( text.substr(start_pos, end_pos - start_pos) );
                     text = text.substr(text.indexOf( ">", end_pos )+1);
                 };
             };

             end_pos = text.search( /<[^>]*$/ );
             if ( -1 == end_pos ) {
                 html.push( text );
             } else {
                 script_buffer = text.substr(    end_pos );
                 html.push(      text.substr( 0, end_pos ) );
             };
//             console.log(html, script_buffer);

         };
         document.writeln = function(html) { write_inner(html + "\n"); };

         if ( arg.apply ) {
             while ( ++i < arguments.length ) args[i-1] = arguments[i];
             arg.apply( this, args );
         } else {
             document.write( arg );
         };

         document.write = document_write;
         clearTimeout( timeout );

         if ( '' != ( html = html.join('') ) ) {
             this.append( html );
         };

         return this;
     };

})(jQuery);


jquery.docwrite.1.0.0.js
/*!
 * docwrite jQuery plugin v1.0.0
 *
 * Copyright 2011, Slando Ltd. <dev@slando.com>
 * Dual licensed under the MIT or GPL Version 2 licenses.
 * http://jquery.org/license
 *
 * Date: Thu Feb 24 11:32:37 GMT 2011
 */

/**
 * @fileOverview
 * Make document.write() a little bit safer
 *
 * // Redirect document.write() to an element:
 * $(element).docwrite( function(arg1, arg2) {
 *     document.write("appended to element");
 *     document.write("<script>document.write('also appended to element')</script>");
 * }, "passed to function as arg1", "passed to function as arg2" );
 *
 * $(document).load(function() {
 *     document.write("will be appended to an invisible <DIV>");
 * });
 *
 * $(document).bind( 'beforedocwrite', function(event, data) {
 *     data.html.replace("whatever was going to be written",
 *                       "what you would rather write");
 *     data.target = $(destination);
 *     data.unused = "pass data to the afterdocwrite event";
 *     event.preventDefault(); // cancel writing
 * });
 *
 * $(document).bind( 'afterdocwrite', function(event, data) {
 *     if ( data.unused ) {
 *         # ...
 *     };
 * });
 */
(function($) {

     var match_script     = /<script ([^>]*)>/i,
         match_script_src = /^<script ([^>]* )?src=/i,
         match_script_end = /<\/script>/i,
         document_write   = document.write,
         write_target     = undefined;

     // document.write() implementation before the ready event has fired:
     function write(html) {

         var data  = { target: null, html: html },
             event = $.Event("beforedocwrite");

         $(document).trigger( event, data );

         if ( !event.isDefaultPrevented() ) {

             if ( null === data.target )
                 document_write.call( document, data.html );
             else if ( data.target.docwrite )
             data.target .docwrite( data.html );
             else
                 $(data.target).docwrite( data.html );

             $(document).trigger( 'afterdocwrite', data );

         }

     };
     function writeln(html) { write( html + "\n" ); };

     // document.write() implementation after the ready event has fired:
     function write_late(html) {

         if ( write_target === undefined ) {
             write_target = document.createElement('DIV');
             write_target.style.display = "none";
             document.body.appendChild(write_target);
             write_target = $(write_target);
         };

         var data  = { target: write_target, html: html },
             event = $.Event("beforedocwrite");

         $(document).trigger( event, data );

         if ( !event.isDefaultPrevented() ) {
             data.target.docwrite( data.html );
             $(document).trigger( 'afterdocwrite', data );
         };

     };
     function writeln_late(html) { write_late( html + "\n" ); };

     document.write   = write;
     document.writeln = writeln;
     $(document).ready(function() {
         document.write   = write_late;
         document.writeln = writeln_late;
     });

     $.fn.docwrite = function(arg) {

         var html = [],
             // we're about to override the system document.write (again)
             // but first we make sure it will be reset, even if we throw an exception:
             document_write   = document.write,
             document_writeln = document.writeln,
             timeout = setTimeout(function() {
                 document.write   = document_write;
                 document.writeln = document_writeln;
             }, 0),
             i=0, args=[];

         /*
          * Protect against nested document.write's:
          *
          *     document.write('<script type="text/javascript">document.write("...")</script>');
          *
          * Browsers ignore <script>s inserted with innerHTML,
          * so we have to parse out and eval() the nested <script>s.
          *
          */
         document.write = function write_inner(text) {

             var start_pos,
                 end_pos; // needs to be a local variable because this is called recursively

             while ( -1 != ( start_pos = text.search(match_script) ) ) {
                 // <script...>

                 html.push( text.substr(0, start_pos) );

                 if ( match_script_src.test( text ) ) {
                     // <script src="...">: remove the "<script" and continue parsing:
                     html.push( text.substr(0, start_pos + 7) );
                     text     = text.substr(   start_pos + 7);
                 } else if ( -1 == ( end_pos = text.search(match_script_end) ) ) {
                     // <script> with no </script>:
                     html.push( text );
                     return;
                 } else {
                     // <script>...</script>, may do a document.write() of its own:
                     start_pos = text.indexOf( '>', start_pos ) + 1;
                     $.globalEval( text.substr(start_pos, end_pos - start_pos) );
                     text = text.substr(end_pos + 9);
                 };
             };

             html.push( text );
         };
         document.writeln = function(html) { write_inner(html + "\n"); };

         if ( arg.apply ) {
             while ( ++i < arguments.length ) args[i-1] = arguments[i];
             arg.apply( this, args );
         } else {
             document.write( arg );
         };

         document.write = document_write;
         clearTimeout( timeout );

         if ( '' != ( html = html.join('') ) ) {
             this.append( html );
         };

         return this;
     };

})(jQuery);


jquery.docwrite.1.1.0.js
/*!
 * docwrite jQuery plugin v1.0.0
 *
 * Copyright 2011, Slando Ltd. <dev@slando.com>
 * Dual licensed under the MIT or GPL Version 2 licenses.
 * http://jquery.org/license
 *
 * Date: Thu Feb 24 11:32:37 GMT 2011
 */

/*!
 * docwrite jQuery plugin v1.1.0
 *
 * Modified by m8rge
 * Added ability to process splitted nesteds scripts
 *
 * Examples:
 * $(element).docwrite("<script type='text/javascript' ");
 * $(element).docwrite("src='alert.js'></scr"+"ipt>");
 *
 * $(element).docwrite("<div>Appended to element</div>" + "<script type='text/javascript'>" + "m='<div>nested '; document.write(m+'");
 * $(element).docwrite("document.write()s are handled");
 * $(element).docwrite(" correctly</div>');" + "</scr"+"ipt>");
 * $(element).docwrite('nezhdanchik');
 *
 * Date: May 04 18:00 GMT+5 2011
 */

/**
 * @fileOverview
 * Make document.write() a little bit safer
 *
 * // Redirect document.write() to an element:
 * $(element).docwrite( function(arg1, arg2) {
 *     document.write("appended to element");
 *     document.write("<script>document.write('also appended to element')</script>");
 * }, "passed to function as arg1", "passed to function as arg2" );
 *
 * $(document).load(function() {
 *     document.write("will be appended to an invisible <DIV>");
 * });
 *
 * $(document).bind( 'beforedocwrite', function(event, data) {
 *     data.html.replace("whatever was going to be written",
 *                       "what you would rather write");
 *     data.target = $(destination);
 *     data.unused = "pass data to the afterdocwrite event";
 *     event.preventDefault(); // cancel writing
 * });
 *
 * $(document).bind( 'afterdocwrite', function(event, data) {
 *     if ( data.unused ) {
 *         # ...
 *     };
 * });
 */
(function($) {

     var match_script     = /<script /i,
         match_script_src = /^<script ([^>]* )?src=/i,
         match_script_end = /<\/script>/i,
         document_write   = document.write,
         write_target     = undefined,
		 script_buffer    = [];

     // document.write() implementation before the ready event has fired:
     function write(html) {

         var data  = { target: null, html: html },
             event = $.Event("beforedocwrite");

         $(document).trigger( event, data );

         if ( !event.isDefaultPrevented() ) {

             if ( null === data.target )
                 document_write.call( document, data.html );
             else if ( data.target.docwrite )
             data.target .docwrite( data.html );
             else
                 $(data.target).docwrite( data.html );

             $(document).trigger( 'afterdocwrite', data );

         }

     };
     function writeln(html) { write( html + "\n" ); };

     // document.write() implementation after the ready event has fired:
     function write_late(html) {

         if ( write_target === undefined ) {
             write_target = document.createElement('DIV');
             write_target.style.display = "none";
             document.body.appendChild(write_target);
             write_target = $(write_target);
         };

         var data  = { target: write_target, html: html },
             event = $.Event("beforedocwrite");

         $(document).trigger( event, data );

         if ( !event.isDefaultPrevented() ) {
             data.target.docwrite( data.html );
             $(document).trigger( 'afterdocwrite', data );
         };

     };
     function writeln_late(html) { write_late( html + "\n" ); };

     document.write   = write;
     document.writeln = writeln;
     $(document).ready(function() {
         document.write   = write_late;
         document.writeln = writeln_late;
     });

     $.fn.docwrite = function(arg) {

         var html = [],
             // we're about to override the system document.write (again)
             // but first we make sure it will be reset, even if we throw an exception:
             document_write   = document.write,
             document_writeln = document.writeln,
             timeout = setTimeout(function() {
                 document.write   = document_write;
                 document.writeln = document_writeln;
             }, 0),
             i=0, args=[];

         /*
          * Protect against nested document.write's:
          *
          *     document.write('<script type="text/javascript">document.write("...")</script>');
          *
          * Browsers ignore <script>s inserted with innerHTML,
          * so we have to parse out and eval() the nested <script>s.
          *
          */
         document.write = function write_inner(text) {
//			 console.log("have text:",text);
             var start_pos,
                 end_pos; // needs to be a local variable because this is called recursively

             while ( -1 != ( start_pos = text.search(match_script) ) ) {
                 // <script...>

                 html.push( text.substr(0, start_pos) );

                 if ( match_script_src.test( text ) ) {
                     // <script src="...">: remove the "<script" and continue parsing:
					 script_buffer.push( text.substr(0, start_pos + 7) );
                     text     = text.substr(   start_pos + 7);
                 } else if ( -1 == ( end_pos = text.search(match_script_end) ) ) {
                     // <script> with no </script>:
					 var tmp = text.substr(start_pos);
					 start_pos = tmp.indexOf( '>', start_pos ) + 1;
					 script_buffer.push(tmp.substr(start_pos));
//					 console.log(html, script_buffer);
                     return;
                 } else {
                     // <script>...</script>, may do a document.write() of its own:
                     start_pos = text.indexOf( '>', start_pos ) + 1;
                     $.globalEval( text.substr(start_pos, end_pos - start_pos) );
                     text = text.substr(end_pos + 9);
                 }
             }
			 if ( script_buffer.length != 0 &&
					 -1 != ( end_pos = text.search(match_script_end) )) {
				 // script buffer have full script
//				 console.log('have script');

				 script_buffer.push(text.substr(0, end_pos)); // append last segment
				 if (match_script_src.test( script_buffer.join('')+text.substr(end_pos, end_pos+9) )) {
				 	 // we have <script src=, so just insert as text
					 script_buffer.push(text.substr(end_pos, end_pos+9)); // append </script> tag
					 html = html.concat(script_buffer);
					 script_buffer = [];
				 } else {
				    // we have <script tag, eval it
					var local_script = script_buffer.join('');
				 	script_buffer = [];
				 	$.globalEval(local_script);
				 }
				 text = text.substr(end_pos + 9);
			 }

			 if (script_buffer.length != 0)
			 	script_buffer.push(text);
			 else
			 	html.push( text );
//			 console.log(html, script_buffer);
         };
         document.writeln = function(html) { write_inner(html + "\n"); };

         if ( arg.apply ) {
             while ( ++i < arguments.length ) args[i-1] = arguments[i];
             arg.apply( this, args );
         } else {
             document.write( arg );
         };

         document.write = document_write;
         clearTimeout( timeout );

         if ( '' != ( html = html.join('') ) ) {
             this.append( html );
         };

         return this;
     };

})(jQuery);
